ntoskrnl/io/iomgr/deviface.c

```diff
@@ -304,7 +305,191 @@ IoGetDeviceInterfaceAlias(IN PUNICODE_STRING SymbolicLinkName,
+    if (!RtlPrefixUnicodeString(&DosDevicesPrefix1, SymbolicLinkName, FALSE) &&
+        !RtlPrefixUnicodeString(&DosDevicesPrefix2, SymbolicLinkName, FALSE))
+    {
+        DPRINT1("IoSetDeviceInterfaceState() invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    /* Make a version without the prefix for further processing */
+    ASSERT(DosDevicesPrefix1.Length == DosDevicesPrefix2.Length);
+    ASSERT(SymbolicLinkName->Length >= DosDevicesPrefix1.Length);
+    LinkNameNoPrefix.Buffer = SymbolicLinkName->Buffer + DosDevicesPrefix1.Length / sizeof(WCHAR);
+    LinkNameNoPrefix.Length = SymbolicLinkName->Length - DosDevicesPrefix1.Length;
+    LinkNameNoPrefix.MaximumLength = LinkNameNoPrefix.Length;
+
+    /* Find the reference string, if any */
+    for (i = 0; i < LinkNameNoPrefix.Length / sizeof(WCHAR); i++)
+    {
+        if (LinkNameNoPrefix.Buffer[i] == L'\\')
+        {
+            break;
+        }
+    }
+    ReferenceStringOffset = i * sizeof(WCHAR);
+
+    /* The GUID is before the reference string or at the end */
+    ASSERT(LinkNameNoPrefix.Length >= ReferenceStringOffset);
+    if (ReferenceStringOffset < GUID_STRING_BYTES + sizeof(WCHAR))
+    {
+        DPRINT1("IoSetDeviceInterfaceState() invalid link name '%wZ'\n", SymbolicLinkName);
+        return STATUS_INVALID_PARAMETER;
+    }
+
+    GuidString.Buffer = LinkNameNoPrefix.Buffer + (ReferenceStringOffset - GUID_STRING_BYTES) / sizeof(WCHAR);
+    GuidString.Length = GUID_STRING_BYTES;
+    GuidString.MaximumLength = GuidString.Length;
+
+    /* Get the device and reference strings from symbolic link by the same way */
+    // TODO: Add some more sanity checks
+    ReferenceString.Buffer = LinkNameNoPrefix.Buffer + ReferenceStringOffset / sizeof(WCHAR);
+    ReferenceString.Length = LinkNameNoPrefix.Length - ReferenceStringOffset;
+    ReferenceString.MaximumLength = ReferenceString.Length;
+
+    // FIXME: Calculation seems to be wrong here
+    DeviceString.Buffer = LinkNameNoPrefix.Buffer + (GuidString.Length - ReferenceString.Length) / sizeof(WCHAR);
+    DeviceString.Length = LinkNameNoPrefix.Length - GuidString.Length - ReferenceString.Length;
+    DeviceString.MaximumLength = DeviceString.Length;
+
+    DPRINT1("Device string is '%wZ'\n", &DeviceString);
+    DPRINT1("GUID string is '%wZ'\n", &GuidString);
+    DPRINT1("Reference string is '%wZ'\n", &ReferenceString);
+
+    /* Convert alias GUID to a string */
+    Status = RtlStringFromGUID(AliasInterfaceClassGuid, &AliasGuidString);
+    if (!NT_SUCCESS(Status))
+    {
+        DPRINT1("RtlStringFromGUID() failed with status 0x%08lx\n", Status);
+        return Status;
+    }
+
+    DPRINT1("Alias GUID is '%wZ'\n", &AliasGuidString);
+
+    /* Build up new symolic link with alias GUID */
+    SymbolicLink.Length = 0;
+    SymbolicLink.MaximumLength = DosDevicesPrefix1.Length + DeviceString.Length + AliasGuidString.Length + ReferenceString.Length;
```

Comment:
I probably also need to do some additional arithmetic operations with `DeviceString.Length` here.